Add support for wildcards in known_hosts #1001

philusky opened the pull request:
Added support for two wildcards: '?' and '*' in known_hosts as well as negation: '!' - same as in sshd.

ploxiln commented on paramiko/hostkeys.py:
```diff
@@ -196,14 +197,34 @@ def _hostname_matches(self, hostname, entry):
+
+        wildcard = wildcard.lstrip("!")
+        wildcard = re.escape(wildcard)
+        wildcard = wildcard.replace("\\?", ".?")
```
`.?` means "0 or 1 of any character" but what you want here is just `.` "exactly one of any character"

ploxiln commented on paramiko/hostkeys.py:
```diff
@@ -196,14 +197,34 @@ def _hostname_matches(self, hostname, entry):
+        if wildcard.startswith("!"):
+            wildcard_negation = True
+
+        wildcard = wildcard.lstrip("!")
```
It would be more efficient to put this in the above if body and just slice off the first char:

```python
if wildcard.startswith("!"):
    wildcard_negation = True
    wildcard = wildcard[1:]
```

ploxiln commented on paramiko/hostkeys.py:
```diff
@@ -196,14 +197,34 @@ def _hostname_matches(self, hostname, entry):
+        wildcard = wildcard.replace("\\?", ".?")
+        wildcard = wildcard.replace("\\*", ".*")
+
+        return re.match(wildcard, hostname) and wildcard_negation is not True
```
It is simpler to write `and not` instead of `is not True`. But we see here that this can't really support negation properly - it just disables the pattern completely. You might as well `return False` above for `if wildcard.startswith("!")`. To support negation properly you would have to make a match here cause all other hostnames/patterns on the same line to also fail to match. For that, you would need a whole different way to parse and handle `known_hosts` in paramiko.

philusky commented:
I really got lost in git ;/ I squashed all commits into one, but some merges happened in the way and there are more than just my changes now. Any tip how to fix that?
Anyhow, I changed `.?` to `.`, removed lstrip() and actually was able to implement proper negation thanks to the fact that `HostKeyEntry` stores all host names / patterns from a single line in `known_hosts`. I wonder if the exception approach is a good one.

ploxiln commented:
That does look like it would work. I'm not sure if `LookupError` is the best exception to use here, maybe a custom one should be created for this.
